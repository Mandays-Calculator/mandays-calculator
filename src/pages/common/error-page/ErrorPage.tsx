import type { ReactElement, ReactNode } from "react";

import { Typography } from "@mui/material";
import { useNavigate } from "react-router-dom";
import { useTranslation } from "react-i18next";

import LocalizationKey from "~/i18n/key";

import PageNotFound from "~/assets/img/page_not_found.png";
import SomethingWentWrong from "~/assets/img/something_wrong.png";

import { CustomButton } from "~/components/form/button";
import { StyledChildContainer, StyledContainer } from ".";
import { Footer } from "~/components";

interface ErrorPageProps {
  type:
    | "something-went-wrong"
    | "not-found"
    | "permission-error"
    | "permission-denied"
    | "configuration-error"
    | "development-mode"
    | "expired-link";
}

const ErrorPage = (props: ErrorPageProps): ReactElement => {
  const { type } = props;
  const navigate = useNavigate();
  const { t } = useTranslation();

  const { common } = LocalizationKey;
  const goBack = (): void => {
    navigate(-1);
  };

  const renderDetails = (): ReactNode => {
    switch (type) {
      case "not-found":
        return (
          <>
            <Typography variant="h5" fontWeight="bold">
              {t(common.pageNotFoundTitle)}
            </Typography>
            <Typography variant="body2" fontWeight="bold">
              {t(common.pageNotFoundDesc)}
            </Typography>
            <CustomButton type="button" onClick={goBack}>
              {t(common.goBackHomeBtnLabel)}
            </CustomButton>
          </>
        );
      case "permission-denied":
        return (
          <>
            <Typography variant="h5" fontWeight="bold">
              {t(common.permissionDeniedTitle)}
            </Typography>
            <Typography variant="body2" fontWeight="bold">
              {t(common.permissionDeniedDesc)}
            </Typography>
            <CustomButton type="button" onClick={goBack}>
              {t(common.goBackHomeBtnLabel)}
            </CustomButton>
          </>
        );
      case "permission-error":
        return (
          <>
            <Typography variant="h5" fontWeight="bold">
              {t(common.permissionErrorTitle)}
            </Typography>
            <Typography variant="body2" fontWeight="bold">
              {t(common.permissionErrorDesc)}
            </Typography>
            <CustomButton type="button" onClick={() => console.log("logout")}>
              {t(common.logout)}
            </CustomButton>
          </>
        );
      case "configuration-error":
        return (
          <>
            <Typography variant="h5" fontWeight="bold">
              {t(common.somethingWentWrongTitle)}
            </Typography>
            <Typography variant="body2" fontWeight="bold">
              {t(common.configurationErrorDesc)}
            </Typography>
          </>
        );
      case "development-mode":
        return (
          <>
            <Typography variant="h5" fontWeight="bold">
              {t(common.pageInProgressTitle)}
            </Typography>
            <Typography variant="body2" fontWeight="bold">
              {t(common.pageInProgressDesc)}
            </Typography>
          </>
        );
      case "expired-link":
        return (
          <>
            <Typography variant="h5" fontWeight="bold">
              Shared Link Expired!
            </Typography>
            <Typography variant="body2" fontWeight="bold">
              Your link has expired. Try requesting from administrator again.
            </Typography>
            <CustomButton type="button" onClick={() => navigate("./login")}>
              {t(common.okayBtn)}
            </CustomButton>
          </>
        );
      default:
        return (
          <>
            <Typography variant="h5" fontWeight="bold">
              {t(common.somethingWentWrongTitle)}
            </Typography>
            <Typography variant="body2" fontWeight="bold">
              {t(common.somethingWentWrongDesc)}
            </Typography>
            <CustomButton type="button" onClick={goBack}>
              {t(common.goBackHomeBtnLabel)}
            </CustomButton>
          </>
        );
    }
  };

  return (
    <StyledContainer $type={type}>
      <StyledChildContainer>
        <img
          src={type === "not-found" ? PageNotFound : SomethingWentWrong}
          alt="error"
          width={330}
          height={280}
        />
      </StyledChildContainer>
      <StyledChildContainer>{renderDetails()}</StyledChildContainer>
      {type === "configuration-error" && <Footer />}
    </StyledContainer>
  );
};

export default ErrorPage;
