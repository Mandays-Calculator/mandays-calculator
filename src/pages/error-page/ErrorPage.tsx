import type { ReactElement, ReactNode } from "react";

import { Typography, styled } from "@mui/material";

import PageNotFound from "~/assets/img/page_not_found.png";
import SomethingWentWrong from "~/assets/img/something_wrong.png";

import { CustomButton } from "~/components/form/button";
import { useNavigate } from "react-router-dom";

const StyledContainer = styled("div")(() => ({
  display: "flex",
  padding: "2rem",
  flexDirection: "column",
  justifyContent: "center",
  alignItems: "center",
  gap: "1rem",
  flex: "1 0 0",
  alignSelf: "stretch",
}));

const StyledChildContainer = styled("div")(() => ({
  display: "flex",
  width: "27rem",
  paddingLeft: "0px",
  flexDirection: "column",
  alignItems: "center",
  gap: "1.375rem",
  textAlign: "center",
}));

interface ErrorPageProps {
  type: "something-went-wrong" | "not-found";
}

const ErrorPage = (props: ErrorPageProps): ReactElement => {
  const { type } = props;
  const navigate = useNavigate();

  const goBack = (): void => {
    navigate(-1);
  };

  const renderDetails = (): ReactNode => {
    if (type === "not-found") {
      return (
        <>
          <Typography
            variant="h5"
            fontWeight="bold"
          >
            PAGE NOT FOUND
          </Typography>
          <Typography
            variant="body2"
            fontWeight="bold"
          >
            Couldn't find that. Try something else or go back home.
          </Typography>
          <CustomButton
            type="button"
            onClick={goBack}
          >
            GO BACK HOME
          </CustomButton>
        </>
      );
    }

    if (type === "something-went-wrong") {
      return (
        <>
          <Typography
            variant="h5"
            fontWeight="bold"
          >
            SOMETHING WENT WRONG
          </Typography>
          <Typography
            variant="body2"
            fontWeight="bold"
          >
            Relax, you didn't break the internet. You may try refreshing the page.
          </Typography>
          <CustomButton
            type="button"
            onClick={goBack}
          >
            GO BACK HOME
          </CustomButton>
        </>
      );
    }

    return null;
  };
  return (
    <StyledContainer>
      <StyledChildContainer>
        <img
          src={type === "not-found" ? PageNotFound : SomethingWentWrong}
          alt="error"
          width={330}
          height={280}
        />
      </StyledChildContainer>
      <StyledChildContainer>{renderDetails()}</StyledChildContainer>
    </StyledContainer>
  );
};

export default ErrorPage;
