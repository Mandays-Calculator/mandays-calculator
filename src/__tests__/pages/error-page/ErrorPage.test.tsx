import { fireEvent, render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import ErrorPage from "~/pages/error-page";

jest.mock("~/assets/img/page_not_found.png", () => "page_not_found.png");
jest.mock("~/assets/img/something_wrong.png", () => "something_wrong.png");
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: jest.fn(),
}));

describe("Error Page Component", () => {
  it("renders the page without breaking", () => {
    const { asFragment, getByText } = render(<ErrorPage type="not-found" />);

    expect(getByText("PAGE NOT FOUND")).toBeInTheDocument();
    expect(getByText("The page you're looking for does not exist.")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders something went wrong", () => {
    const { asFragment, getByText } = render(<ErrorPage type="something-went-wrong" />);
    expect(getByText("SOMETHING WENT WRONG")).toBeInTheDocument();
    expect(
      getByText(
        "Please contact the administrator. You can send a report at this email sample@email.com"
      )
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("clicked go back button", () => {
    const navigate = jest.fn();
    jest.spyOn(require("react-router-dom"), "useNavigate").mockReturnValue(navigate);

    const { getByText } = render(
      <Router>
        <ErrorPage type="not-found" />
      </Router>
    );

    fireEvent.click(getByText("GO BACK HOME"));
    expect(navigate).toHaveBeenCalledWith(-1);
  });
});
