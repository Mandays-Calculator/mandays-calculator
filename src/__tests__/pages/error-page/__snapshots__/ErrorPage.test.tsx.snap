// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Page Component renders something went wrong 1`] = `
<DocumentFragment>
  <div
    class="css-5xmk2m"
  >
    <div
      class="css-108k6cn"
    >
      <img
        alt="error"
        height="280"
        src="something_wrong.png"
        width="330"
      />
    </div>
    <div
      class="css-108k6cn"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-bezmlo-MuiTypography-root"
      >
        common.somethingWentWrong.title
      </h5>
      <p
        class="MuiTypography-root MuiTypography-body2 css-1d3njat-MuiTypography-root"
      >
        common.somethingWentWrong.description
      </p>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-u89a6x-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        common.button.goBackHome
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Error Page Component renders the page without breaking 1`] = `
<DocumentFragment>
  <div
    class="css-5xmk2m"
  >
    <div
      class="css-108k6cn"
    >
      <img
        alt="error"
        height="280"
        src="something_wrong.png"
        width="330"
      />
    </div>
    <div
      class="css-108k6cn"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-bezmlo-MuiTypography-root"
      >
        common.notFound.title
      </h5>
      <p
        class="MuiTypography-root MuiTypography-body2 css-1d3njat-MuiTypography-root"
      >
        common.notFound.description
      </p>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-u89a6x-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        common.button.goBackHome
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
